share/spack/gitlab/cloud_e4s_pipelines/.gitlab-ci.yml

```diff
@@ -1,60 +1,116 @@
-stages: [generate, build]
+stages: [ "generate", "build" ]
+
+default:
+  image: { "name": "ghcr.io/scottwittenburg/ecpe4s-ubuntu18.04-runner-x86_64:2020-09-01", "entrypoint": [""] }
+
+########################################
+# Job templates
+########################################
+
+.pr:
+  only:
+  - /^github\/pr[\d]+_.*$/
+  variables:
+    SPACK_PR_BRANCH: ${CI_COMMIT_REF_NAME}
+    SPACK_IS_PR_PIPELINE: "True"
+
+.develop:
+  only:
+  - /^github\/develop$/
+  variables:
+    SPACK_IS_PR_PIPELINE: "False"
 
 .generate:
   stage: generate
   script:
     - . "./share/spack/setup-env.sh"
     - spack --version
-    - cd share/spack/gitlab/cloud_e4s_pipelines/stacks/e4s
+    - cd share/spack/gitlab/cloud_e4s_pipelines/stacks/${SPACK_CI_STACK_NAME}
     - spack env activate --without-view .
     - spack ci generate --check-index-only
-      --output-file "${CI_PROJECT_DIR}/jobs_scratch_dir/e4s_pipeline.yml"
+      --output-file "${CI_PROJECT_DIR}/jobs_scratch_dir/cloud-ci-pipeline.yml"
   artifacts:
     paths:
-      - "${CI_PROJECT_DIR}/jobs_scratch_dir/e4s_pipeline.yml"
+      - "${CI_PROJECT_DIR}/jobs_scratch_dir/cloud-ci-pipeline.yml"
   tags: ["spack", "public", "medium", "x86_64"]
   interruptible: true
 
+.pr-generate:
+  extends: [ ".pr", ".generate" ]
+
+.develop-generate:
+  extends: [ ".develop", ".generate" ]
+
+.build:
+  stage: build
+
+.pr-build:
+  extends: [ ".pr", ".build" ]
+  variables:
+    AWS_ACCESS_KEY_ID: ${PR_MIRRORS_AWS_ACCESS_KEY_ID}
+    AWS_SECRET_ACCESS_KEY: ${PR_MIRRORS_AWS_SECRET_ACCESS_KEY}
+
+.develop-build:
+  extends: [ ".develop", ".build" ]
+  variables:
+    AWS_ACCESS_KEY_ID: ${MIRRORS_AWS_ACCESS_KEY_ID}
+    AWS_SECRET_ACCESS_KEY: ${MIRRORS_AWS_SECRET_ACCESS_KEY}
+    SPACK_SIGNING_KEY: ${PACKAGE_SIGNING_KEY}
+
+########################################
+# TEMPLATE FOR ADDING ANOTHER PIPELINE
+########################################
+# .my-super-cool-stack:
+#   variables:
+#     SPACK_CI_STACK_NAME: my-super-cool-stack
+#
```

Comment:
This is so nice!  Does it make sense to add an extra doc line here reminding developers to put their stack into a directory of the same name?  e.g.:

```
# Add super cool stack to spack repo:
# 
#     share/spack/gitlab/gitlab_cloud_pipelines/stacks/${SPACK_CI_STACK_NAME}/spack.yaml
```

Reply:
Certainly